src/main/scala/mesosphere/marathon/core/instance/Instance.scala

```diff
@@ -32,18 +32,56 @@ case class Instance(
+      InstancePhase.Launching
+    } else if (maybeHighestPriorityState.exists(s => s == Condition.Running || s == Condition.Unreachable || s == Condition.Killing)) {
+      InstancePhase.Active
+    } else if (state.goal == Goal.Running) {
+      InstancePhase.Scheduled
+    } else {
+      InstancePhase.Terminal
+    }
+  }
+
+  /**
+    * This picks
+    * Previously [[Condition]] was used as a status for both tasks and instances.
+    * We moved away from using that for instances but it's still needed for backward compatibility on API layer.
+    * @param now current time
+    * @return computed [[Condition]] as a summary of task conditions
+    */
+  @deprecated("This should be used only in the API layer for backward compatibility and internally inside instance")
+  def summarizedTaskStatus(now: Timestamp): Condition = {
+    InstanceState.conditionFromTasks(tasksMap.values, now, unreachableStrategy).getOrElse(Condition.Unknown)
+  }
+
   val runSpecId: PathId = instanceId.runSpecId
 
   // An instance has to be considered as Reserved if at least one of its tasks is Reserved.
-  def isReserved: Boolean = tasksMap.values.exists(_.status.condition == Condition.Reserved)
-
-  def isReservedTerminal: Boolean = tasksMap.values.exists(_.isReservedTerminal)
+  lazy val isReserved: Boolean = tasksMap.valuesIterator.exists(_.status.condition == Condition.Reserved)
+  lazy val isReservedTerminal: Boolean = tasksMap.valuesIterator.exists(_.isReservedTerminal)
+
+  // these state cannot change without the instance itself being changed and it's immutable so we are safe to use lazy val
+  lazy val isStaging: Boolean = summarizedTaskStatus(Timestamp.now()) == Condition.Staging
+  lazy val isStarting: Boolean = summarizedTaskStatus(Timestamp.now()) == Condition.Starting
+  lazy val isKilling: Boolean = summarizedTaskStatus(Timestamp.now()) == Condition.Killing
+  lazy val isRunning: Boolean = summarizedTaskStatus(Timestamp.now()) == Condition.Running
+
+  // these states can change depending on current time even without this Instance being changed
+  def isUnreachable(now: Timestamp): Boolean = summarizedTaskStatus(now) == Condition.Unreachable
+  def isUnreachableInactive(now: Timestamp): Boolean = unreachableStrategy match {
+    case _: UnreachableEnabled => summarizedTaskStatus(now) == Condition.UnreachableInactive
+    case _ => false
+  }
 
-  def isKilling: Boolean = state.condition == Condition.Killing
-  def isRunning: Boolean = state.condition == Condition.Running
-  def isUnreachable: Boolean = state.condition == Condition.Unreachable
-  def isUnreachableInactive: Boolean = state.condition == Condition.UnreachableInactive
-  def isActive: Boolean = state.condition.isActive
+  def isTerminal: Boolean = phase(Timestamp.now()) == InstancePhase.Terminal
+  def isActive: Boolean = {
+    val instancePhase = phase(Timestamp.now())
+    instancePhase == InstancePhase.Active || instancePhase == InstancePhase.Launching
```

Comment:
This seems odd to me. We have an active phase but `isActive` checks if it is launching. How about `isActiveOrLaunching`?